src/main/java/com/matt/forgehax/mods/BreadCrumbs.java

```diff
@@ -0,0 +1,385 @@
+    for (Anchor it : unvisited) distances.put(it, Double.POSITIVE_INFINITY);
+    distances.put(root, 0.D);
+    final Map<Anchor, Anchor> prev = new HashMap<>();
+
+    while (!unvisited.isEmpty()) {
+      final Anchor u = Collections.min(unvisited, Comparator.comparingDouble(distances::get));
+      unvisited.remove(u);
+      if (u == target) break;
+
+      for (Anchor v : u.connected) {
+        final double alt = distances.get(u) + length(u, v);
+        if (alt < distances.get(v)) {
+          distances.put(v, alt);
+          prev.put(v, u);
+        }
+      }
+    }
+
+    return getPath(root, target, prev);
+  }
+
+  private static <T> Optional<T> getLast(List<T> list) {
+    return !list.isEmpty() ? Optional.of(list.get(list.size() - 1)) : Optional.empty();
+  }
+
+  @SubscribeEvent
+  public void onPlayerUpdate(LocalPlayerUpdateEvent event) {
+    if (Helper.getModManager().get(FreecamMod.class).map(ToggleMod::isEnabled).orElse(false)) return;
+
+    final Vec3d playerPos = new Vec3d(MC.player.posX, MC.player.posY, MC.player.posZ);
+
+    // first tick of a trail
+    if (this.rootAnchor == null) {
+      this.rootAnchor = new Anchor(playerPos);
+      this.newestAnchor = rootAnchor;
+      this.dimension = MC.player.dimension;
+      return;
+    }
+
+    final Set<Anchor> visibleAnchors = getVisibleAnchors(this.rootAnchor);
+    final Set<Anchor> noLongerVisible = Sets.difference(this.visibleLastTick, visibleAnchors);
+
+    final Set<Anchor> allOldAnchors = Sets.difference(getAllAnchors(rootAnchor), Collections.singleton(this.newestAnchor));
+
+    final Optional<Anchor> closest = !allOldAnchors.isEmpty() ? // TODO: make this a small list?
+        Optional.of(Collections.min(allOldAnchors, Comparator.comparingDouble(anch -> anch.pos.distanceTo(playerPos))))
+        : Optional.empty();
+
+    if (noLongerVisible.contains(this.newestAnchor) || closest.map(noLongerVisible::contains).orElse(false)) { // new anchor
+      final Anchor newAnchor = new Anchor(playerPos);
+      Stream.of(noLongerVisible, visibleAnchors)
+          .flatMap(Set::stream)
+          .distinct()
+          .forEach(anch -> anch.connectAnchor(newAnchor));
+      this.newestAnchor = newAnchor;
+    } else { // new point for the last anchor
+      final List<Vec3d> points = this.newestAnchor.points;
+      final Optional<Vec3d> prev = getLast(points);
+      // don't spam points if we don't move
+      if (!prev.isPresent() || prev.get().distanceTo(playerPos) > 0.01) {
+        points.add(playerPos);
+      }
+    }
+
+    this.pointsToDraw = getPointsToDraw();
+
+    this.visibleLastTick = getVisibleAnchors(this.rootAnchor); // should probably just reuse the list we just made
+  }
+
+  private static <T> List<T> partialList(List<T> list, int smoothness) {
+    if (smoothness == 1) return list;
+    // TODO: return view of list
+    final List<T> out = new ArrayList<>((int)Math.ceil(list.size() / (double)smoothness));
```

Comment:
Smoothness is the wrong term, it impies that increasing it will increase the smoothness of data points, not reduce them. "Roughness"?

Also epic int float maths

Also with the loop below it will occasionally leave out the last data point for s>1 (might not look pretty)

Reply:
yes
yes
no